src/GafferScene/BranchCreator.cpp

```diff
@@ -140,38 +164,44 @@ class BranchCreator::BranchesData : public IECore::Data
 				const auto parent = branchCreator->parentPlugPath();
 				if( parent )
 				{
+					ScenePlug::PathScope pathScope( context, *parent );
 					hashBranch( branchCreator, *parent, h );
 				}
 			}
 		}
 
-		const IECore::PathMatcher &parentPaths() const
+		const IECore::PathMatcher &destinationPaths() const
 		{
-			return m_parentPaths;
+			return m_destinationPaths;
+		}
+
+		// Maps from destination path to a list of source paths.
+		const map<ScenePlug::ScenePath, vector<ScenePlug::ScenePath>> &sourcePaths() const
+		{
+			return m_sourcePaths;
 		}
 
 	private :
 
 		static void hashBranch( const BranchCreator *branchCreator, const ScenePlug::ScenePath &path, IECore::MurmurHash &h )
 		{
-			if( path.size() )
-			{
-				h.append( path.data(), path.size() );
-			}
-			else
-			{
-				h.append( 0 );
-			}
+			branchCreator->destinationPlug()->hash( h );
 		}
 
 		void addBranch( const BranchCreator *branchCreator, const ScenePlug::ScenePath &path )
 		{
+			ScenePlug::ScenePath destination;
+			ScenePlug::stringToPath( branchCreator->destinationPlug()->getValue(), destination );
+
 			tbb::spin_mutex::scoped_lock lock( m_mutex );
-			m_parentPaths.addPath( path );
+			m_destinationPaths.addPath( destination );
+			m_sourcePaths[destination].push_back( path );
 		}
 
 		tbb::spin_mutex m_mutex;
-		IECore::PathMatcher m_parentPaths;
+		IECore::PathMatcher m_destinationPaths;
+
+		map<ScenePlug::ScenePath, vector<ScenePlug::ScenePath>> m_sourcePaths;
```

Comment:
Using ScenePath as a key in a map means repeated comparisons between vectors?  That seems like a possible perf hazard?  I guess a theoretically ideal solution would be allowing storage at each node in a PathMatcher ... oh, glancing at the final code, that is where you end up.

Reply:
Yeah, sorry, this is the change-of-data-structures-midway-through-the-work that I mentioned. I should have jumped to the tree of locations in the original refactor commit, but I hadn't realised I needed it at that point.